executor/join.go

```diff
@@ -374,22 +379,20 @@ func (e *HashJoinExec) Next() (*Row, error) {
 			return nil, errors.Trace(err)
 		}
 	}
-	var (
-		row *Row
-		err error
-		ok  bool
-	)
-	select {
-	case row, ok = <-e.resultRows:
-	case err, ok = <-e.resultErr:
-	}
-	if err != nil {
-		e.finished.Store(true)
-		return nil, errors.Trace(err)
-	}
-	if !ok {
-		return nil, nil
+	if e.cursor >= len(e.rows) {
+		result, ok := <-e.resultCh
+		if !ok {
+			return nil, nil
+		}
+		if result.err != nil {
+			e.finished.Store(true)
+			return nil, errors.Trace(result.err)
+		}
+		e.rows = result.rows
```

Comment:
If rows is of zero length, it would panic on line 394.

Even though after reviewing the code, I can make sure the length of rows will never be zero, but that takes effort, and code may change in the future, adding a check on length of result.rows doesn't hurt.

The same applies to UnionExec